File: src/logic/pieces/rook.py
from logic.move import Move
from .piece import Piece

class Rook(Piece):
    def legal_moves(self, board: "Board", / , looking_for_check = False) -> list[Move]:
        ret = []

        # looking east
        ret.extend(self._look_direction(board, 1, 0))

        # looking south
        ret.extend(self._look_direction(board, 0, -1))

        # looking west
        ret.extend(self._look_direction(board, -1, 0))

        # looking north
        ret.extend(self._look_direction(board, 0, 1))

        if not looking_for_check and board.is_check_for(self.colour):
            return self.keep_only_blocking(ret, board)

        return ret

File: cpp/src/pieces/rook.cpp
#include "../board.hpp"
#include "../coords.hpp"
#include "../move.hpp"
#include "piece.hpp"

#include <vector>

std::vector<Move> rook_moves(const Board& b, const Coords xy) {
    std::vector<Move> ret;
    auto e = look_direction(b, xy, 1, 0);
    ret.insert(ret.end(), e.begin(), e.end());

    auto s = look_direction(b, xy, 0, -1);
    ret.insert(ret.end(), s.begin(), s.end());

    auto w = look_direction(b, xy, -1, 0);
    ret.insert(ret.end(), w.begin(), w.end());

    auto n = look_direction(b, xy, 0, 1);
    ret.insert(ret.end(), n.begin(), n.end());

    return ret;
}
